### ISSUE TITLE
FCOS train with Pascal VOC

### ISSUE THREAD

**Issue by TWDH** [open-mmlab/mmdetection#3339]
Hello @hellock , I'd like to train pascal voc using FCOS, I tried to modify the config file myself. Firstly, I added ```'../_base_/datasets/voc0712.py'``` in the _base_ and replace the num_classes from 80 to 20 in bbox_head, btw, I train it on only 1 GPU.  All the things I followed the FCOS and PascalVOC settings but the result is extremely low about 25.2% mAP(VOC dataset). would you please give some suggestion on where should I modify? Here is the full config file, Thanks!
```
_base_ = [
    '../_base_/datasets/voc0712.py',
    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
]
# model settings
model = dict(
    type='FCOS',
    pretrained='open-mmlab://detectron/resnet50_caffe',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=False),
        norm_eval=True,
        style='caffe'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=1,
        add_extra_convs=True,
        extra_convs_on_inputs=False,  # use P5
        num_outs=5,
        relu_before_extra_convs=True),
    bbox_head=dict(
        type='FCOSHead',
        num_classes=20,
        in_channels=256,
        stacked_convs=4,
        feat_channels=256,
        strides=[8, 16, 32, 64, 128],
        loss_cls=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),
        loss_bbox=dict(type='IoULoss', loss_weight=1.0),
        loss_centerness=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0)))

# training and testing settings
train_cfg = dict(
    assigner=dict(
        type='MaxIoUAssigner',
        pos_iou_thr=0.5,
        neg_iou_thr=0.4,
        min_pos_iou=0,
        ignore_iof_thr=-1),
    allowed_border=-1,
    pos_weight=-1,
    debug=False)
test_cfg = dict(
    nms_pre=1000,
    min_bbox_size=0,
    score_thr=0.05,
    nms=dict(type='nms', iou_thr=0.5),
    max_per_img=100)

img_norm_cfg = dict(
    mean=[102.9801, 115.9465, 122.7717], std=[1.0, 1.0, 1.0], to_rgb=False)

train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1333, 800), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1333, 800),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    train=dict(pipeline=train_pipeline),
    val=dict(pipeline=test_pipeline),
    test=dict(pipeline=test_pipeline))
# optimizer
optimizer = dict(type='SGD',lr=0.01/2, paramwise_cfg=dict(bias_lr_mult=2., bias_decay_mult=0.))

optimizer_config = dict(
    _delete_=True, grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='constant',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[3])
total_epochs = 4

```

**Comment by gjxgjxgjxgjx** [comment]
I also encountered this problem. I set the learning rate to 0.002, which seemed to solve the problem.

**Comment by wyman123** [comment]
I have met the same problem, have you solved it?

**Comment by LRQ516** [comment]
> Hello @hellock , I'd like to train pascal voc using FCOS, I tried to modify the config file myself. Firstly, I added `'../_base_/datasets/voc0712.py'` in the _base_ and replace the num_classes from 80 to 20 in bbox_head, btw, I train it on only 1 GPU. All the things I followed the FCOS and PascalVOC settings but the result is extremely low about 25.2% mAP(VOC dataset). would you please give some suggestion on where should I modify? Here is the full config file, Thanks!
> 
> ```
> _base_ = [
>     '../_base_/datasets/voc0712.py',
>     '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
> ]
> # model settings
> model = dict(
>     type='FCOS',
>     pretrained='open-mmlab://detectron/resnet50_caffe',
>     backbone=dict(
>         type='ResNet',
>         depth=50,
>         num_stages=4,
>         out_indices=(0, 1, 2, 3),
>         frozen_stages=1,
>         norm_cfg=dict(type='BN', requires_grad=False),
>         norm_eval=True,
>         style='caffe'),
>     neck=dict(
>         type='FPN',
>         in_channels=[256, 512, 1024, 2048],
>         out_channels=256,
>         start_level=1,
>         add_extra_convs=True,
>         extra_convs_on_inputs=False,  # use P5
>         num_outs=5,
>         relu_before_extra_convs=True),
>     bbox_head=dict(
>         type='FCOSHead',
>         num_classes=20,
>         in_channels=256,
>         stacked_convs=4,
>         feat_channels=256,
>         strides=[8, 16, 32, 64, 128],
>         loss_cls=dict(
>             type='FocalLoss',
>             use_sigmoid=True,
>             gamma=2.0,
>             alpha=0.25,
>             loss_weight=1.0),
>         loss_bbox=dict(type='IoULoss', loss_weight=1.0),
>         loss_centerness=dict(
>             type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0)))
> 
> # training and testing settings
> train_cfg = dict(
>     assigner=dict(
>         type='MaxIoUAssigner',
>         pos_iou_thr=0.5,
>         neg_iou_thr=0.4,
>         min_pos_iou=0,
>         ignore_iof_thr=-1),
>     allowed_border=-1,
>     pos_weight=-1,
>     debug=False)
> test_cfg = dict(
>     nms_pre=1000,
>     min_bbox_size=0,
>     score_thr=0.05,
>     nms=dict(type='nms', iou_thr=0.5),
>     max_per_img=100)
> 
> img_norm_cfg = dict(
>     mean=[102.9801, 115.9465, 122.7717], std=[1.0, 1.0, 1.0], to_rgb=False)
> 
> train_pipeline = [
>     dict(type='LoadImageFromFile'),
>     dict(type='LoadAnnotations', with_bbox=True),
>     dict(type='Resize', img_scale=(1333, 800), keep_ratio=True),
>     dict(type='RandomFlip', flip_ratio=0.5),
>     dict(type='Normalize', **img_norm_cfg),
>     dict(type='Pad', size_divisor=32),
>     dict(type='DefaultFormatBundle'),
>     dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
> ]
> test_pipeline = [
>     dict(type='LoadImageFromFile'),
>     dict(
>         type='MultiScaleFlipAug',
>         img_scale=(1333, 800),
>         flip=False,
>         transforms=[
>             dict(type='Resize', keep_ratio=True),
>             dict(type='RandomFlip'),
>             dict(type='Normalize', **img_norm_cfg),
>             dict(type='Pad', size_divisor=32),
>             dict(type='ImageToTensor', keys=['img']),
>             dict(type='Collect', keys=['img']),
>         ])
> ]
> data = dict(
>     samples_per_gpu=4,
>     workers_per_gpu=4,
>     train=dict(pipeline=train_pipeline),
>     val=dict(pipeline=test_pipeline),
>     test=dict(pipeline=test_pipeline))
> # optimizer
> optimizer = dict(type='SGD',lr=0.01/2, paramwise_cfg=dict(bias_lr_mult=2., bias_decay_mult=0.))
> 
> optimizer_config = dict(
>     _delete_=True, grad_clip=dict(max_norm=35, norm_type=2))
> # learning policy
> lr_config = dict(
>     policy='step',
>     warmup='constant',
>     warmup_iters=500,
>     warmup_ratio=1.0 / 3,
>     step=[3])
> total_epochs = 4
> ```

Hello, I also encountered the same problem, how did you solve it? Looking forward to your reply.

**Comment by xandercai** [comment]
Same issue. Why it was closed? Any update?

**Comment by xandercai** [comment]
Just for the record.

1. "open-mmlab://detectron/resnet50_caffe" did not fit for current resnet50 model, should change to "torchvision://resnet50"
2. no need to overwrite "../_base_/datasets/voc0712.py" (pipeline something).

**Comment by jacksonsc007** [comment]
@TWDH 
Hi, I checked your full training log and found the reason:
In the final config as shown in the full training log, "data.train" contains two types of pipelines: the one in "data.train.dataset.pipeline" and the other in "data.train.pipeline". If you take a closer look your could notice the image_scale in the Resize operation of them is different, i.e. "img_scale=(1000, 600)" and "img_scale=(1333, 800)". 
According to my experience, the input image size followed "data.train.dataset.pipeline" and is resized to (1000,600), which is contradictory against the size during validation and test stage (Your could check data.val and data.test).
I think the problem exsited in the training with VOC dataset is due to the dataset type, i.e. RepeatDataset, with another "pipeline" key.
